tools/syz-trace2syz/parser/intermediate_types.go

```diff
@@ -0,0 +1,531 @@
+// Copyright 2018 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package parser
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/prog"
+	"github.com/google/syzkaller/tools/syz-trace2syz/config"
+)
+
+type operation int
+
+const (
+	orOp       = iota // OR = |
+	andOp             // AND = &
+	xorOp             // XOR = ^
+	lshiftOp          // LSHIFT = <<
+	rshiftOp          // RSHIFT = >>
+	onescompOp        // ONESCOMP = ~
+	timesOp           // TIMES = *
+	landOp            // LAND = &&
+	lorOp             // LOR = ||
+	lequalOp          // LEQUAL = ==
+	negOp             // MINUS -x
+	plusOp            // A + B
+	minusOp           // A - B
+
+	exprTypeName    string = "Expression Type"
+	groupTypeName   string = "Group Type"
+	bufferTypeName  string = "Buffer Type"
+	pointerTypeName string = "Pointer Type"
+)
+
+// TraceTree struct contains intermediate representation of trace
+// If a trace is multiprocess it constructs a trace for each type
+type TraceTree struct {
+	TraceMap map[int64]*Trace
+	Ptree    map[int64][]int64
+	RootPid  int64
+	Filename string
+}
+
+// NewTraceTree initializes a TraceTree
+func NewTraceTree() (tree *TraceTree) {
+	tree = &TraceTree{
+		TraceMap: make(map[int64]*Trace),
+		Ptree:    make(map[int64][]int64),
+		RootPid:  -1,
+	}
+	return
+}
+
+func (tree *TraceTree) contains(pid int64) bool {
+	if _, ok := tree.TraceMap[pid]; ok {
+		return true
+	}
+	return false
+}
+
+func (tree *TraceTree) add(call *Syscall) *Syscall {
+	if tree.RootPid < 0 {
+		tree.RootPid = call.Pid
+	}
+	if !call.Resumed {
+		if !tree.contains(call.Pid) {
```

Comment:
This is the only call to contains, so I would just inline it here and remove contains function. Also it's not necessary to make an empty slice, nil value is a perfectly valid nil slice. So this can be just:
```
if tree.TraceMap[pid] == nil {
    tree.TraceMap[pid] = newTrace()
}
```